### PR TITLE
Deprecated public constructor of FSTCompiler in favor of the Builder.

### PR DESCRIPTION
**Pull request by cavorite**
### Description

Addresses issue #12695: use the builder for `FSTCompiler` as the only way to create instances of that class. We're replacing the old invocations of the constructor to use the new method, and marking it as deprecated (so that it's no longer public in an upcoming release).

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/util/fst/FSTCompiler.java [hunk 1]
```diff
@@ -122,8 +122,11 @@ public class FSTCompiler<T> {
   /**
    * Instantiates an FST/FSA builder with default settings and pruning options turned off. For more
    * tuning and tweaking, see {@link Builder}.
+   *
+   * @deprecated Use FSTCompiler.Builder instead. This method will be private in an upcoming major
```
Comment (dungba88) [inline lucene/core/src/java/org/apache/lucene/util/fst/FSTCompiler.java:126]: I think we would probably remove it instead of making private?

**Comment by dungba88** [comment]
Thank you for change!

File: lucene/core/src/java/org/apache/lucene/util/fst/FSTCompiler.java (hunk 1 above)
Comment (cavorite) [reply lucene/core/src/java/org/apache/lucene/util/fst/FSTCompiler.java]: Good point. I updated the Javadoc and mentioned that in the `CHANGES.txt` file.

**Comment by dungba88** [comment]
@mikemccand can you help to review this?

**Review by dungba88 — APPROVED** [review]

File: lucene/CHANGES.txt [hunk 2]
```diff
@@ -7,6 +7,8 @@ http://s.apache.org/luceneversions
 
 API Changes
 ---------------------
+* GITHUB-12695: Deprecated public constructor of FSTCompiler. Please use FSTCompiler.Builder
```
Comment (mikemccand) [inline lucene/CHANGES.txt:10]: Note that the FST API is `@lucene.internal` so we are free to 1) skip deprecation -- just remove the old API, and 2) backport to 9.9.

We are delaying backporting all these recent FST changes for now (letting things bake in trunk) so we would first commit this to `main` and later backport with the full patch.  So for now, let's leave this `CHANGES` entry in 10.0.0 section, but let's remove the deprecated APIs entirely?

File: lucene/core/src/java/org/apache/lucene/util/fst/FSTCompiler.java [hunk 3]
```diff
@@ -125,8 +125,11 @@ public class FSTCompiler<T> {
   /**
    * Instantiates an FST/FSA builder with default settings and pruning options turned off. For more
    * tuning and tweaking, see {@link Builder}.
+   *
+   * @deprecated Use FSTCompiler.Builder instead. This method will be removed in an upcoming major
+   *     release.
    */
-  // TODO: remove this?  Builder API should be the only entry point?
+  @Deprecated
   public FSTCompiler(FST.INPUT_TYPE inputType, Outputs<T> outputs) {
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/util/fst/FSTCompiler.java:133]: Let's remove instead?

**Review by mikemccand — COMMENTED** [review]
Thank you @cavorite -- this looks great!  I just think we should be more aggressive and remove (not deprecate) the API since it is `@lucene.experimental`.

File: lucene/core/src/java/org/apache/lucene/util/fst/FSTCompiler.java (hunk 3 above)
Comment (mikemccand) [reply lucene/core/src/java/org/apache/lucene/util/fst/FSTCompiler.java:133]: I'll merge this now, and we can separately fold this feedback.  It's holding up some changes on #12624.

**Comment by mikemccand** [comment]
Thank you @cavorite!  Much cleaner to use a consistent API for building FSTs...

File: lucene/CHANGES.txt (hunk 2 above)
Comment (cavorite) [reply lucene/CHANGES.txt:10]: Thanks for the clarification, @mikemccand. I'll post another PR removing the public constructor.

**Comment by dungba88** [comment]
Thanks @cavorite , I have incorporated this change to #12624 . Removing the constructor would also be great as it means there is less thing needs to be backward compatible :)

**Pull request merged**